diff --git a/src/condor_c++_util/store_cred.cpp b/src/condor_c++_util/store_cred.cpp
index 9d43fc8..49da998 100644
--- a/src/condor_c++_util/store_cred.cpp
+++ b/src/condor_c++_util/store_cred.cpp
@@ -170,7 +170,7 @@ char* getStoredCredential(const char *username, const char *domain)
 
 int store_cred_service(const char *user, const char *pw, int mode)
 {
-	char *at = strchr(user, '@');
+	const char *at = strchr(user, '@');
 	if ((at == NULL) || (at == user)) {
 		dprintf(D_ALWAYS, "store_cred: malformed user name\n");
 		return FAILURE;
