diff -ru condor-7.0.2.orig/src/configure.ac condor-7.0.2/src/configure.ac
--- condor-7.0.2.orig/src/configure.ac	2008-06-11 10:04:54.000000000 -0500
+++ condor-7.0.2/src/configure.ac	2008-06-11 10:04:07.000000000 -0500
@@ -877,6 +877,12 @@
     CompilerMinor="2"
     CompilerPatch="1"
     ;;
+ "4.3.2" )
+    CompilerKind="GCC"
+    CompilerMajor="4"
+    CompilerMinor="3"
+    CompilerPatch="2"
+    ;;
  * )
     if test "x$enable_gcc_version_check" = "xyes"; then
       AC_MSG_RESULT([ERROR])
