diff --git a/src/configure.ac b/src/configure.ac
index af0150f..1f3c575 100644
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -2258,6 +2258,9 @@ if test "x$_cv_libc_kind" = "xGLIBC" -a \( "x$_cv_arch" = "xI386" -o "x$_cv_arch
   "2.7." ) # F8 uses glibc 2.7-2
     including_glibc_ext=NO
     ;;
+  "2.7.90" ) # F9
+    including_glibc_ext=NO
+    ;;
   * )
     AC_MSG_CHECKING(glibc)
     AC_MSG_RESULT([ERROR])
