--- a/src/condor_examples//condor_config.generic.orig	2011-05-17 08:19:40.000000000 -0500
+++ b/src/condor_examples/condor_config.generic	2011-05-17 08:26:48.000000000 -0500
@@ -77,7 +77,7 @@
 ##  Where is the local condor directory for each host?  
 ##  This is where the local config file(s), logs and
 ##  spool/execute directories are located
-LOCAL_DIR		= $(TILDE)
+LOCAL_DIR		= /var/lib/condor
 #LOCAL_DIR		= $(RELEASE_DIR)/hosts/$(HOSTNAME)
 
 ## Looking for LOCAL_CONFIG_FILE? You will not find it here. Instead
