diff --git a/config/configure.cf.in b/config/configure.cf.in
index 78ba6ff..6fa1b8f 100644
--- a/config/configure.cf.in
+++ b/config/configure.cf.in
@@ -132,8 +132,10 @@ VENDOR_CPLUS_NAME = @_cv_vendor_cplus_name@
 CONFIGURE_DEFS = @DEFS@
 CONFIGURE_LIBS = @LIBS@
 CONFIGURE_CFLAGS = @CFLAGS@
-CONFIGURE_CPPFLAGS = @CPPFLAGS@
+CONFIGURE_GSOAP_CFLAGS = @GSOAP_CFLAGS@
+CONFIGURE_CPPFLAGS = @CPPFLAGS@ @GSOAP_CFLAGS@
 CONFIGURE_LDFLAGS = @LDFLAGS@
+CONFIGURE_GSOAP_LDFLAGS = @GSOAP_LDFLAGS@
 
 LEX = @LEX@
 LEXLIB = @LEXLIB@
diff --git a/src/aclocal.m4 b/src/aclocal.m4
index 1930aff..11b2cb8 100644
--- a/src/aclocal.m4
+++ b/src/aclocal.m4
@@ -220,6 +220,8 @@ AC_DEFUN([MF_LIB_CHECK],
   [AS_IF([test "x$_dir" != x],
     [_ldflags_save="$LDFLAGS"
      _cflags_save="$CFLAGS"
+     _ldflags="-L$_dir/lib"
+     _cflags="-I$_dir/include"
      LDFLAGS="$LDFLAGS -L$_dir/lib"
      CFLAGS="$CFLAGS -I$_dir/include"])
      _libs=
@@ -233,8 +235,13 @@ AC_DEFUN([MF_LIB_CHECK],
         LDFLAGS=$_ldflags_save
         CFLAGS=$_cflags_save
         AC_MSG_WARN([$1: could not find $_failure])],
-       [cv_ext_]m4_tolower($1)[=yes
-        LIBS="$LIBS $_libs"])])
+        m4_if(x$3, x,
+         [LIBS="$LIBS $_libs"],
+         m4_toupper($3)[_LDFLAGS="$_ldflags $_libs"]
+         m4_toupper($3)[_CFLAGS="$_cflags"]
+         [LDFLAGS="$_ldflags_save"
+          CFLAGS="$_cflags_save"])
+        [cv_ext_]m4_tolower($1)[=yes])])
 
 
 #######################################################################
diff --git a/src/condor_c-gahp/Imakefile b/src/condor_c-gahp/Imakefile
index 5151566..b43db4c 100644
--- a/src/condor_c-gahp/Imakefile
+++ b/src/condor_c-gahp/Imakefile
@@ -8,7 +8,7 @@ all_target( $(PROGRAMS) )
 CFLAGS = $(STD_C_FLAGS)
 
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(cgahp)
 
diff --git a/src/condor_collector.V6/Imakefile b/src/condor_collector.V6/Imakefile
index 624bfb6..b0ed290 100644
--- a/src/condor_collector.V6/Imakefile
+++ b/src/condor_collector.V6/Imakefile
@@ -3,7 +3,7 @@ all_target(collector_engine.a $(NAME))
 
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS)
-LIB = collector_engine.a $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = collector_engine.a $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 ENGINE_OBJ = hashkey.o collector_engine.o
 
diff --git a/src/condor_daemon_core.V6/Imakefile b/src/condor_daemon_core.V6/Imakefile
index 48230b9..09fb8b4 100644
--- a/src/condor_daemon_core.V6/Imakefile
+++ b/src/condor_daemon_core.V6/Imakefile
@@ -24,7 +24,7 @@ GSOAP_OBJS = soap_core.o httpget.o
 
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS) $(FIX_BROKEN_GCC) $(PURIFY_FLAG) $(GSOAP_FLAGS)
-LIB = $(STD_LIBS)
+LIB = $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 OBJ = $(GSOAP_OBJS) timer_manager.o daemon_core.o daemon_core_main.o \
   daemon_core_instantiate.o condor_ipverify.o condor_perms.o \
diff --git a/src/condor_dagman/Imakefile b/src/condor_dagman/Imakefile
index ff49359..a2a0361 100644
--- a/src/condor_dagman/Imakefile
+++ b/src/condor_dagman/Imakefile
@@ -12,7 +12,7 @@ C_PLUS_FLAGS = -I. $(STD_C_PLUS_FLAGS) \
 	-Wall
 
 CFLAGS = $(STD_C_FLAGS)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS) New_classads_lib
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS) New_classads_lib
 TOOL_LIB = $(STD_LIBS) New_classads_lib
 
 soapobjs(dagman)
diff --git a/src/condor_dbmsd/Imakefile b/src/condor_dbmsd/Imakefile
index 4005cb9..9386872 100644
--- a/src/condor_dbmsd/Imakefile
+++ b/src/condor_dbmsd/Imakefile
@@ -3,7 +3,7 @@ all_target( $(NAME) )
 
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS) $(TT_INC)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(TT_LIB)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(TT_LIB) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(dbmsd)
 
diff --git a/src/condor_dcskel/Imakefile b/src/condor_dcskel/Imakefile
index 348de93..ad6c2de 100644
--- a/src/condor_dcskel/Imakefile
+++ b/src/condor_dcskel/Imakefile
@@ -3,7 +3,7 @@ all_target( $(NAME) )
 
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(dcskel)
 
diff --git a/src/condor_eventd.V2/Imakefile b/src/condor_eventd.V2/Imakefile
index 78b24ae..32a4364 100644
--- a/src/condor_eventd.V2/Imakefile
+++ b/src/condor_eventd.V2/Imakefile
@@ -3,7 +3,7 @@ all_target( $(NAME) )
 
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS) 
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(Eventd)
 
diff --git a/src/condor_gridmanager/Imakefile b/src/condor_gridmanager/Imakefile
index cbc9e76..3a71d3a 100644
--- a/src/condor_gridmanager/Imakefile
+++ b/src/condor_gridmanager/Imakefile
@@ -12,7 +12,7 @@ ORACLE_JOB_LIB = /p/condor/workspaces/jfrey/oracle/OraHome1/lib/libclient9.a /p/
 
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(ORACLE_JOB_FLAGS) $(STD_C_PLUS_FLAGS)
-LIB = $(ORACLE_JOB_LIB) $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(ORACLE_JOB_LIB) $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(gridmanager)
 
diff --git a/src/condor_had/Imakefile b/src/condor_had/Imakefile
index 88dd166..f22a03a 100644
--- a/src/condor_had/Imakefile
+++ b/src/condor_had/Imakefile
@@ -3,7 +3,7 @@ all_target($(HAD_NAME))
 
 CFLAGS = $(STD_C_FLAGS) -D_POSIX_PTHREAD_SEMANTICS
 C_PLUS_FLAGS =  $(STD_C_PLUS_FLAGS) -D_POSIX_PTHREAD_SEMANTICS
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 HAD_OBJ =  StateMachine.o HAD.o Utils.o $(SOAP_OBJS)
 #HAD_OBJ =  StateMachine.o HAD.o Utils.o soap_hadC.o soap_hadServer.o soap_hadStub.o
diff --git a/src/condor_master.V6/Imakefile b/src/condor_master.V6/Imakefile
index 00dd2e5..f6e2b74 100644
--- a/src/condor_master.V6/Imakefile
+++ b/src/condor_master.V6/Imakefile
@@ -3,7 +3,7 @@ all_target($(NAME))
 
 CFLAGS = $(STD_C_FLAGS) 
 C_PLUS_FLAGS =  $(STD_C_PLUS_FLAGS)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(master)
 
diff --git a/src/condor_negotiator.V6/Imakefile b/src/condor_negotiator.V6/Imakefile
index c3c6d5e..73a253e 100644
--- a/src/condor_negotiator.V6/Imakefile
+++ b/src/condor_negotiator.V6/Imakefile
@@ -3,7 +3,7 @@ all_target($(NAME))
 
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS)
-LIB = $(DAEMONCORE_LIB) $(NETMAN_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(NETMAN_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(negotiator)
 
diff --git a/src/condor_schedd.V6/Imakefile b/src/condor_schedd.V6/Imakefile
index dc24674..31eb8a1 100644
--- a/src/condor_schedd.V6/Imakefile
+++ b/src/condor_schedd.V6/Imakefile
@@ -4,7 +4,7 @@ all_target(libqmgmt.a $(NAME))
 C_PLUS_FLAGS = -I../condor_daemon_core.V6 $(STD_C_PLUS_FLAGS)
 CFLAGS =  -I../condor_daemon_core.V6 $(STD_C_FLAGS)
 
-LIB= $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB= $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 QMGMT_OBJ = qmgmt.o qmgmt_receivers.o qmgmt_common.o schedd_files.o
 QMGMT_LIB_OBJ = qmgmt_send_stubs.o qmgr_lib_support.o qmgmt_common.o \
diff --git a/src/condor_shadow.V6.1/Imakefile b/src/condor_shadow.V6.1/Imakefile
index df1f4cd..5a546a3 100644
--- a/src/condor_shadow.V6.1/Imakefile
+++ b/src/condor_shadow.V6.1/Imakefile
@@ -4,7 +4,7 @@ all_target($(NAME))
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS) -I../condor_schedd.V6
 
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(shadow)
 
diff --git a/src/condor_startd.V6/Imakefile b/src/condor_startd.V6/Imakefile
index 8086cb2..833125e 100644
--- a/src/condor_startd.V6/Imakefile
+++ b/src/condor_startd.V6/Imakefile
@@ -18,7 +18,7 @@ BACKFILL_OBJS = $(BOINC_OBJS) backfill_mgr.o
 
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS)
 
-LIB= $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB= $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(startd)
 
diff --git a/src/condor_starter.V6.1/Imakefile b/src/condor_starter.V6.1/Imakefile
index b060a35..008ed11 100644
--- a/src/condor_starter.V6.1/Imakefile
+++ b/src/condor_starter.V6.1/Imakefile
@@ -10,7 +10,7 @@ all_target($(NAME) $(RSH_NAME) condor_gridshell)
 CFLAGS = $(STD_C_FLAGS) -I../condor_schedd.V6
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS) -I../condor_schedd.V6
 
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 RSH_LIB = $(STD_LIBS) $(CLASSAD_LIB)
 
 #if IS_SGI_IRIX62
diff --git a/src/condor_transferd/Imakefile b/src/condor_transferd/Imakefile
index 09d6b4e..662770c 100644
--- a/src/condor_transferd/Imakefile
+++ b/src/condor_transferd/Imakefile
@@ -3,7 +3,7 @@ all_target( $(NAME) )
 
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(transferd)
 
diff --git a/src/condor_tt/Imakefile b/src/condor_tt/Imakefile
index 074cc8b..307f94c 100644
--- a/src/condor_tt/Imakefile
+++ b/src/condor_tt/Imakefile
@@ -6,7 +6,7 @@ all_target($(NAME))
 CFLAGS = $(STD_C_FLAGS)
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS) $(TT_INC)
 LDFLAGS = $(STD_LDFLAGS) $(LIBCRYPT)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(TT_LIB) ../condor_quill/libquill.a
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(TT_LIB) $(CONFIGURE_GSOAP_LDFLAGS) ../condor_quill/libquill.a
 
 OBJ = tt_main.o ttmanager.o jobqueuecollection.o jobqueuedbmanager.o condor_ttdb.o
 
diff --git a/src/condor_vm-gahp/Imakefile b/src/condor_vm-gahp/Imakefile
index 995f950..97fc34c 100644
--- a/src/condor_vm-gahp/Imakefile
+++ b/src/condor_vm-gahp/Imakefile
@@ -7,7 +7,7 @@ all_target( $(PROGRAMS) condor_vm_xen.sh condor_vm_vmware.pl)
 CFLAGS = $(STD_C_FLAGS)
 
 C_PLUS_FLAGS = $(STD_C_PLUS_FLAGS)
-LIB = $(DAEMONCORE_LIB) $(STD_LIBS)
+LIB = $(DAEMONCORE_LIB) $(STD_LIBS) $(CONFIGURE_GSOAP_LDFLAGS)
 
 soapobjs(vmgahp)
 
diff --git a/src/configure.ac b/src/configure.ac
index 5bd1643..9fc253d 100644
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -2434,11 +2434,13 @@ CHECK_EXTERNAL([gsoap], [2.7.6c-p2], [soft],
                 LDFLAGS="$LDFLAGS -Wl[,]--unresolved-symbols=ignore-in-shared-libs"
                 AS_IF([test "x$cv_ext_globus" = xyes -o "x$cv_ext_openssl" = xyes],
                   [MF_LIB_CHECK([gsoap],
-                                [[gsoapssl++ soap_ssl_accept]])],
+                                [[gsoapssl++ soap_ssl_accept]], [gsoap])],
                   [MF_LIB_CHECK([gsoap],
-                                [[gsoap++ soap_accept]])])
+                                [[gsoap++ soap_accept]], [gsoap])])
                 AS_IF([test "x$cv_ext_gsoap" = xno],
                   [LDFLAGS="$_ldflags_save"])])
+AC_SUBST(GSOAP_LDFLAGS, $GSOAP_LDFLAGS)
+AC_SUBST(GSOAP_CFLAGS, $GSOAP_CFLAGS)
 
 # We need to make sure HAVE_OPENSSL_SSL_H is defined otherwise
 # programs will not link properly as stdsoap2.h will not include ssl
